# 0.Introduction

<br>

The database used in this analysis was taken from the [Kaggle](https://www.kaggle.com/) website.

<br>

The following analysis was carried out on the database called ["sales in period walmart"](https://www.kaggle.com/datasets/willianoliveiragibin/sales-in-period-walmart?select=_5wuCXmjSp25ws6zPYB8Rg_462096727ec040c1b1a5c9ab5397acf1_Inventory.csv), in order to understand the sales made by year, item, and store, to find out if there are, and what are the differences between seasons, locations, or products sold.

<br>

## index

1. Importing Pandas
2. EDA
3. Data Cleaning
4. Transforming Variables
5. Creating hypotheses
6. Combining variables for analysis

# 1.Importing Pandas

In [2]:
import pandas as pd

#### 1.1 Reading Wallmart's monthly sales csv file

In [3]:
pd.read_csv('Monthly_Sales_2017_2020.csv')

SalesId  StoreId  ProductId        Date  UnitPrice  Quantity
0         82319    22726        590  2019-12-02     0.0525        93
1         15022    21754        390  2017-11-19     5.1100        28
2         11624    71053        883  2020-07-13     7.3675        33
3         63101    22914        658  2019-05-12     2.0825        76
4         29702    22623        632  2020-07-20     0.6475         8
...         ...      ...        ...         ...        ...       ...
199995    57660    85123        682  2019-07-01     4.4625        75
199996    18783    22727        656  2017-04-18     1.6800        30
199997    25064    22912        831  2019-06-10     2.5025        60
199998    43683    22622        902  2018-08-06     3.2025        98
199999    44211    22749        775  2018-06-22     1.8550        32

[200000 rows x 6 columns]

#### 1.2 Creating a variable to store the database

In [21]:
walmart_sales = pd.read_csv('Monthly_Sales_2017_2020.csv')

# 2.Exploratory Data Analysis(EDA)

#### 2.1 looking at the first and last 5 lines

In [22]:
display(walmart_sales)

SalesId  StoreId  ProductId        Date  UnitPrice  Quantity
0         82319    22726        590  2019-12-02     0.0525        93
1         15022    21754        390  2017-11-19     5.1100        28
2         11624    71053        883  2020-07-13     7.3675        33
3         63101    22914        658  2019-05-12     2.0825        76
4         29702    22623        632  2020-07-20     0.6475         8
...         ...      ...        ...         ...        ...       ...
199995    57660    85123        682  2019-07-01     4.4625        75
199996    18783    22727        656  2017-04-18     1.6800        30
199997    25064    22912        831  2019-06-10     2.5025        60
199998    43683    22622        902  2018-08-06     3.2025        98
199999    44211    22749        775  2018-06-22     1.8550        32

[200000 rows x 6 columns]

#### 2.2 Statistical information on the product's unit price column (UnitPrice)

In [26]:
walmart_sales.UnitPrice.describe()

count    200000.000000
mean          3.313406
std           2.812360
min           0.017500
25%           1.277500
50%           2.765000
75%           4.287500
max          21.157500
Name: UnitPrice, dtype: float64

#### 2.3 Statistical information on the product quantity sold column (Quantity)

In [27]:
walmart_sales.Quantity.describe()

count    200000.000000
mean         50.383910
std          28.870702
min           1.000000
25%          25.000000
50%          50.000000
75%          75.000000
max         100.000000
Name: Quantity, dtype: float64

#### 2.4 Database size

In [28]:
all_sales.shape

(200000, 6)

# 3.Data Cleaning

#### 3.1 Checking information

<br>

By checking, we can see that we don't have any null information, and we can also see that the 'Date' variable is not in datetime format.

In [23]:
walmart_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   SalesId    200000 non-null  int64  
 1   StoreId    200000 non-null  int64  
 2   ProductId  200000 non-null  int64  
 3   Date       200000 non-null  object 
 4   UnitPrice  200000 non-null  float64
 5   Quantity   200000 non-null  int64  
dtypes: float64(1), int64(4), object(1)
memory usage: 9.2+ MB


#### 3.2 Checking for duplicates

In [24]:
walmart_sales.duplicated().any()

False

# 4.Transforming Variables

#### 4.1 Converting the "Date" column to datetime format

In [42]:
walmart_sales['Date'] = pd.to_datetime(walmart_sales['Date'])

#### 4.2 Viewing the data type again

In [37]:
walmart_sales.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 11007 to 29602
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   SalesId    200000 non-null  int64         
 1   StoreId    200000 non-null  int64         
 2   ProductId  200000 non-null  int64         
 3   Date       200000 non-null  datetime64[ns]
 4   UnitPrice  200000 non-null  float64       
 5   Quantity   200000 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(4)
memory usage: 10.7 MB


#### 4.3 Sorting the "Date" column in ascending order

In [43]:
walmart_sales = walmart_sales.sort_values('Date')
walmart_sales.head()

SalesId  StoreId  ProductId       Date  UnitPrice  Quantity
11007     17024    21777        985 2017-01-01     2.6775        99
40438     57773    22632        839 2017-01-01     2.1700        52
176084    56186    21035        161 2017-01-01     2.2925         3
155090    84975    21755          9 2017-01-01     4.0775        87
56662     45056    22749        609 2017-01-01     6.0900        24

#### 4.4 Filtering the database by "Date"

<br>

The database created (walmart_sales) contains sales from 2017 to 2020.
This complete database has 200,000 rows, which slows down analysis due to the amount of data. 
So I separated the 2017 sales into a new database (sales_2017), with 49,910 rows, thus making the database smaller and consequently making the analysis more targeted and faster.

#### 4.5 Creating a variable to store 2017 sales data

In [44]:
sales_2017 = walmart_sales[walmart_sales['Date'].dt.year == 2017]

In [209]:
# sales_2017 = all_sales[all_sales['Date'].dt.year == 2017]
# sales_2017_jan = all_sales[all_sales['Date'].between('2017-01-01', '2017-01-31')]

#### 4.6 Creating the Total Sales 2017 column (Total_Sales)

In [45]:
sales_2017['Total_Sales'] = sales_2017['UnitPrice'] * sales_2017['Quantity']
sales_2017.head()

C:\Users\Guiga\AppData\Local\Temp\ipykernel_15800\1651390691.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sales_2017['Total_Sales'] = sales_2017['UnitPrice'] * sales_2017['Quantity']


SalesId  StoreId  ProductId       Date  UnitPrice  Quantity  \
11007     17024    21777        985 2017-01-01     2.6775        99   
40438     57773    22632        839 2017-01-01     2.1700        52   
176084    56186    21035        161 2017-01-01     2.2925         3   
155090    84975    21755          9 2017-01-01     4.0775        87   
56662     45056    22749        609 2017-01-01     6.0900        24   

        Total_Sales  
11007      265.0725  
40438      112.8400  
176084       6.8775  
155090     354.7425  
56662      146.1600

# 5.Creating hypotheses

1. Was there a month/quarter/semester of the year that sold more?
    - [ ] If so, what was the difference in sales?
    - [ ] And the reason for this difference?
    - [x] If not, what is the average sales and what is the difference between the period that sold the most and the least?
    - [x] Is it a big difference or are they all average?
2. Was there a product that sold more or made more profit?
    - [x] What are the cheapest products?
    - [x] What are the most expensive products?
    - [x] Which product made the most profit?
    - [x] Which product made the least profit?
    - [x] Which product sold the most?
    - [x] Which product sold the least?
3. Was there a store that sold more or made more profit?
    - [ ] Which store sold the most?
    - [ ] Which store sold the least?
    - [ ] Which store made the most profit?
    - [ ] Which store made the least profit?

# 6.Combining variables for analysis

#### 6.1 Grouping sales by Month in 2017

In [46]:
monthly_sales_2017 = sales_2017.groupby(pd.Grouper(key='Date', freq='M')).agg({'Total_Sales': 'sum'})
display(monthly_sales_2017)

Total_Sales
Date                   
2017-01-31  701304.6600
2017-02-28  634254.5300
2017-03-31  721855.8900
2017-04-30  700876.2425
2017-05-31  712926.7600
2017-06-30  690128.0050
2017-07-31  686223.7900
2017-08-31  686763.3850
2017-09-30  700209.5975
2017-10-31  710114.0725
2017-11-30  693119.1050
2017-12-31  727249.4075

Analyzing monthly sales for the year 2017, we can see that every month has a well-balanced total of sales on average.

1. The month with the lowest sales value in 2017 is February with a value equal to 634254.53.
2. And the month with the highest sales value in 2017 is December with a value equal to 727249.40.

#### 6.2 Discovering the cheapest and most expensive product of 2017

In [55]:
sales_2017.head()

SalesId  StoreId  ProductId       Date  UnitPrice  Quantity  \
11007     17024    21777        985 2017-01-01     2.6775        99   
40438     57773    22632        839 2017-01-01     2.1700        52   
176084    56186    21035        161 2017-01-01     2.2925         3   
155090    84975    21755          9 2017-01-01     4.0775        87   
56662     45056    22749        609 2017-01-01     6.0900        24   

        Total_Sales  
11007      265.0725  
40438      112.8400  
176084       6.8775  
155090     354.7425  
56662      146.1600

#### Sorting by UnitPrice

In [97]:
sales_2017_up = sales_2017.sort_values('UnitPrice')

In [98]:
sales_2017_up.head()

SalesId  StoreId  ProductId       Date  UnitPrice  Quantity  \
28384     86812    22310        245 2017-05-16     0.0175        38   
60640     22910    22310        245 2017-11-30     0.0175        12   
43682     72567    21756        245 2017-05-31     0.0175         2   
175054    40319    22912        245 2017-05-13     0.0175        73   
157385    29536    22752        245 2017-01-30     0.0175        16   

        Total_Sales  
28384        0.6650  
60640        0.2100  
43682        0.0350  
175054       1.2775  
157385       0.2800

In [99]:
unitprice2 = sales_2017_up[["ProductId", "UnitPrice"]].drop_duplicates()
unitprice2.sort_values("UnitPrice")

ProductId  UnitPrice
28384         245     0.0175
160555        471     0.0350
167440        195     0.0350
158528        988     0.0350
111784        671     0.0350
...           ...        ...
39834         367    15.5400
89164          89    16.8350
108166        117    19.9325
24412         248    21.0875
35190         885    21.1575

[1000 rows x 2 columns]

Analyzing the produduct unit prices in 2017, we can conclude that:

1. The cheapest product is "245", costing 0.0175.

2. The most expensive product is the "885", costing 21.1575.

#### 6.3 Discovering the most and least profitable product of the year 2017

#### Grouping by Quantity and Total_Sales and saving it in the variable product_sales_2017

In [47]:
product_sales_2017 = sales_2017.groupby('ProductId').agg({'Quantity': 'sum', 'Total_Sales': 'sum'})

#### Sorting by Total Sales

In [49]:
product_sales_2017_TS = product_sales_2017.sort_values('Total_Sales')

#### The 5 most and least profitable products

In [54]:
display(product_sales_2017_TS)

Quantity  Total_Sales
ProductId                       
245            2158      37.7650
895            2105      73.6750
671            2251      78.7850
988            2346      82.1100
471            2451      85.7850
...             ...          ...
805            3041   44649.4825
885            2165   45805.9875
89             2877   48434.2950
248            2757   58138.2375
117            3049   60774.1925

[1000 rows x 2 columns]

Analyzing the grouping created, we can conclude that:

1. The most profitable product of the year 2017 is “117”, with a total sales value of 60774.1925
2. The least profitable product of the year 2017 is the “245”, with a total sales value of 37.7650

#### 6.4 Discovering the most and least sold product of 2017

#### Sorting by Quantity Sold

In [63]:
product_sales_2017_Q = product_sales_2017.sort_values('Quantity')

In [64]:
display(product_sales_2017_Q)

Quantity  Total_Sales
ProductId                       
318            1375    1540.0000
948            1472    1932.0000
353            1488    7551.6000
235            1501     472.8150
927            1510    4333.7000
...             ...          ...
9              3596   14662.6900
826            3612   22818.8100
586            3749   26439.8225
479            3758    3945.9000
696            4158    9459.4500

[1000 rows x 2 columns]

Analyzing the grouping created, we can conclude that:

1. The best-selling product of 2017 is “696”, with the amount of 4158 total products sold.
2. The least sold product in 2017 is “318”, with the amount of 1375 total products sold.

#### 6.5 Grouping Sales by Store in 2017

In [70]:
store_sales_2017 = sales_2017.groupby('StoreId').agg({'Total_Sales': 'sum'})

In [71]:
store_sales_2017.shape

(34, 1)

#### Ordenando por Vendas por Loja

In [72]:
store_sales_2017 = store_sales_2017.sort_values('Total_Sales')

In [74]:
display(store_sales_2017)

Total_Sales
StoreId             
21791    224123.3400
84029    232017.1000
22726    233693.7750
48187    235589.0600
22912    238157.9725
84406    238609.8750
22727    238618.8175
84969    240660.5425
22310    241711.8725
22622    241862.9150
21883    242046.1575
21754    242253.8300
21756    242447.1525
21035    242975.6000
21724    243122.8975
22728    244310.7975
10002    244897.5375
22633    245154.1575
85123    245720.4575
22745    246020.3550
22752    247007.6525
22960    247053.6600
21730    247978.5875
71053    250631.9375
22623    252504.6125
22913    252853.6150
22748    253212.8375
22914    253355.7425
37444    253653.4350
21755    254209.5500
84879    255107.5625
22632    261807.1925
21777    265439.3000
22749    266215.5475

Analyzing the grouping created, we can conclude that:

1. The store that sold the most in 2017 was '22749', with a total of 266215.5475 sales.
2. The store that sold the least in 2017 was '21791', with a total of 224123.3400 sales.